plaso/parsers/megasync.py

```diff
@@ -0,0 +1,203 @@
+from plaso.containers import time_events
+from plaso.lib import definitions
+from plaso.parsers import logger
+from plaso.parsers import manager
+from plaso.parsers import text_parser
+
+
+class MEGASyncEventData(events.EventData):
+  """MEGASync log event data.
+
+  Attributes:
+    log_level (str): log level.
+    message (str): message.
+  """
+
+  DATA_TYPE = 'megasync:log:line'
+
+  def __init__(self):
+    """Initializes event data."""
+    super(MEGASyncEventData, self).__init__(data_type=self.DATA_TYPE)
+    self.log_level = None
+    self.message = None
+
+class MEGASyncParser(text_parser.PyparsingSingleLineTextParser):
+  """Parses MEGASync log files"""
+
+  NAME = 'megasync'
+  DATA_FORMAT = 'MEGASync log file'
+
+  # Some types of MEGASync log lines can be very long.
+  MAX_LINE_LENGTH = 65536
+
+  _ENCODING = 'utf-8'
+
+  _TWO_DIGITS = text_parser.PyparsingConstants.TWO_DIGITS
+
+  # Timestamp format is: mm/dd-hh:mm:ss.######
+  # For example: 03/21-04:13:44.621454
+  _TIMESTAMP = pyparsing.Group(
+      _TWO_DIGITS.setResultsName('month') + pyparsing.Suppress('/') +
+      _TWO_DIGITS.setResultsName('day') + pyparsing.Suppress('-') +
+      text_parser.PyparsingConstants.TIME_MSEC_ELEMENTS
+  ).setResultsName('timestamp')
+
+  _THREAD_NAME = pyparsing.Word(pyparsing.printables)
+
+  _LOG_LEVEL = (
+      pyparsing.Literal('DBG') |
+      pyparsing.Literal('INFO') |
+      pyparsing.Literal('WARN') |
+      pyparsing.Literal('DTL') |
+      pyparsing.Literal('ERR') |
+      pyparsing.Literal('CRIT')).setResultsName('log_level')
+
+  _MESSAGE = (pyparsing.White(' ' ,min=1,max=2).suppress() +
+      pyparsing.restOfLine().setResultsName('message'))
+
+  _BASE_LOG_LINE = _TIMESTAMP + _THREAD_NAME
+
+  _LOG_LINE = _TIMESTAMP + _THREAD_NAME + _LOG_LEVEL + _MESSAGE
+
+  # Indicates that the last log line was repeated multiple times.
+  _REPEAT_LINE = (
+      pyparsing.Suppress('[repeated x') +
+      text_parser.PyparsingConstants.INTEGER +
+      pyparsing.Suppress("]")
+  ).setResultsName("repeats")
+
+  _PROGRAM_START = pyparsing.Literal(
+      '----------------------------- program start -----------------------------')
+
+  LINE_STRUCTURES = [
+      ('line', _LOG_LINE),
+      ('repeat', _REPEAT_LINE),
+      ('program_start', _PROGRAM_START)]
+
+  _LINES_OF_INTEREST = [
+    'Transfer (UPLOAD) finished',
+    'Transfer (UPLOAD) starting',
+    'Upload complete',
+    'Creating thumb/preview'
+  ]
+
+  def __init__(self):
+    """Initializes a parser."""
+    super(MEGASyncParser, self).__init__()
+    self._last_month = 0
+    self._maximum_year = 0
+    self._year_use = 0
+
+  def _UpdateYear(self, mediator, month):
+    """Updates the year to use for events, based on last observed month.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
+      month (int): month observed by the parser, where January is 1.
+    """
```

Comment:
MEGASync compresses rotated-out log files using the gzip format. the `mtime` timestamp in the gzip stream, however, is empty
<img width="448" alt="image" src="https://user-images.githubusercontent.com/4998544/184106156-c44ffe80-c21f-4583-b9a4-77e123a03027.png">
making this approach not work. (and currently I just set the estimate to the current year in case of epoch timestamps)

I'm struggling a bit to find a good way to estimate the year of the log file in plaso, other than relying on `--preferred-year`. 

I've considered some things like accessing the timestamps of the compressed file itself (should be accurate except for some edge cases, but also don't see that the APIs would allow for this currently).